File: exports/scala/cfg/DataStream.scala
package perfect.cfg

import java.io.File
import java.nio.charset.Charset
import java.nio.file.Files
import java.util.List

object DataStream {
  private val magicStringForNewLine: String = ".g9~/"

  def create(dataFile: String, inputEncoding: String): DataStream = {
    try {
      return new DataStream(Files.readAllLines(new File(dataFile).toPath, Charset.forName(inputEncoding)))
    }
    catch {
      case e: Exception => {
        e.printStackTrace()
        throw new RuntimeException("data file:" + dataFile + " loads fail!")
      }
    }
  }
}

final class DataStream(val lines: List[String]) {
  private var index: Int = 0

  private def getNext: String = {
    if (index < lines.size) lines.get({
      index += 1; index - 1
    })
    else null
  }

  private def error(err: String) {
    throw new RuntimeException(s"$index $err")
  }

  private def getNextAndCheckNotEmpty : String = {
    val s = getNext
    if (s == null) error("read not enough")
    s
  }

  def getBool() : Boolean = {
    val s: String = getNextAndCheckNotEmpty.toLowerCase
    s match {
      case "true" => true
      case "false" => false
      case _ =>
        error(s + " isn't bool")
        false
    }
  }

  def getInt(): Int = {
    val s: String = getNextAndCheckNotEmpty
    s.toInt
  }

  def getLong() : Long = {
    val s: String = getNextAndCheckNotEmpty
    java.lang.Long.parseLong(s)
  }

  def getFloat() : Float = {
    val s: String = getNextAndCheckNotEmpty
    java.lang.Float.parseFloat(s)
  }

  def getString() : String = {
    getNextAndCheckNotEmpty.replace(DataStream.magicStringForNewLine, "\n")
  }

  def getObject(name: String) : CfgObject = {
    try {
      Class.forName(name).getConstructor(classOf[DataStream]).newInstance(this).asInstanceOf[CfgObject]
    }
    catch {
      case e: Exception => throw new RuntimeException(e)
    }
  }
}

File: exports/cs/DataStream.cs
﻿using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;

namespace cfg
{
    public class DataStream
    {
		public const string MagicStringForNewLine = ".g9~/";

        private readonly string[] line;
        private int index;
        DataStream(string file, string encoding)
        {
            line = File.ReadAllLines(file);
            index = 0;
        }

        public string GetNext()
        {
            return index < line.Length ? line[index++] : null;
        }

        void Error(string err)
        {
            throw new Exception(err);
        }

        string GetNextAndCheckNotEmpty()
        {
            var s = GetNext();
            if (s == null)
                Error("read not enough");
            return s;
        }

        public string GetString()
        {
            return GetNextAndCheckNotEmpty().Replace(MagicStringForNewLine, "\n");
        }

        public float GetFloat()
        {
            return float.Parse(GetNextAndCheckNotEmpty());
        }

        public int GetInt()
        {
            return int.Parse(GetNextAndCheckNotEmpty());
        }

        public long GetLong()
        {
            return long.Parse(GetNextAndCheckNotEmpty());
        }
		
        public bool GetBool()
        {
            var s = GetNextAndCheckNotEmpty().ToLower();
            if (s == "true")
                return true;
            if (s == "false")
                return false;
            Error(s + " isn't bool");
            return false;
        }

        public cfg.CfgObject GetObject(string name)
        {
            return (cfg.CfgObject)Type.GetType(name).GetConstructor(new[] { typeof(cfg.DataStream) }).Invoke(new object[] { this });
        }

        public static DataStream Create(string file, string encoding)
        {
            return new DataStream(file, encoding);
        }
		
    }
}
